tools/ramdump/ramdump_tool.c

```diff
@@ -80,69 +88,221 @@ static int b_read(int fd, uint8_t *buf, int size)
-		printf("Receiving address failed, ret = %d\n", ret);
+	/* Receive number of memory regions from TARGET */
+	ret = read(dev_fd, &c, 1);
+	if (ret != 1) {
+		printf("Receiving number of regions failed, ret = %d\n", ret);
 		return -1;
 	}
+	number_of_regions = c;
 
-	ret = b_read(dev_fd, (uint8_t *)&ramdump_size, 4);
-	if (ret != 4) {
-		printf("Receiving size failed, ret = %d\n", ret);
+	/* Array to store memory region address */
+	rd_regionx_start = malloc((number_of_regions + 2) * sizeof(int));
+	if (!rd_regionx_start) {
 		return -1;
 	}
-
-	printf("%s: ramdump_address = %08x, ramdump_size = %d\n", __func__, ramdump_address, ramdump_size);
-
-	snprintf(bin_file, BINFILE_NAME_SIZE, "ramdump_0x%08x_0x%08x.bin", ramdump_address, (ramdump_address + ramdump_size));
-
-	bin_fp = fopen(bin_file, "w");
-	if (bin_fp == NULL) {
-		printf("%s create failed\n", bin_file);
+	/* Array to store memory region size */
+	rd_regionx_size = malloc((number_of_regions + 2) * sizeof(int));
+	if (!rd_regionx_size) {
+		return -1;
+	}
+	/* Array to store heap index corresponding to memory regions */
+	rd_regionx_idx = malloc((number_of_regions + 2) * sizeof(int));
+	if (!rd_regionx_idx) {
+		return -1;
+	}
+	/* Array to map selected regions to be dumped */
+	rd_regionx_map = malloc((number_of_regions + 2) * sizeof(int));
+	if (!rd_regionx_map) {
 		return -1;
 	}
 
-	printf("[>");
-	fflush(stdout);
+	/* Receive memory address, size & heap index for memory regions from TARGET */
+	for (i = 2; i <= number_of_regions + 1; i++) {
+		ret = b_read(dev_fd, (uint8_t *)&mem_address, 4);
+		if (ret != 4) {
+			printf("Receiving address failed, ret = %d\n", ret);
+			return -1;
+		}
+		rd_regionx_start[i] = mem_address;
 
-	while (ramdump_size) {
-		ret = read(dev_fd, &buf, 1);
-		if (ret != 1) {
-			printf("Receiving ramdump %dTH byte failed, ret = %d\n", count, ret);
-			fclose(bin_fp);
+		ret = b_read(dev_fd, (uint8_t *)&mem_size, 4);
+		if (ret != 4) {
+			printf("Receiving size failed, ret = %d\n", ret);
 			return -1;
 		}
+		rd_regionx_size[i] = mem_size;
 
-		ret = fwrite(&buf, 1, 1, bin_fp);
+		ret = read(dev_fd, &c, 1);
 		if (ret != 1) {
-			printf("Writing ramdump %dTH byte failed, ret = %d\n", count, ret);
-			fclose(bin_fp);
+			printf("Receiving size failed, ret = %d\n", ret);
 			return -1;
 		}
+		rd_regionx_idx[i] = c;
+	}
 
-		count++;
-		ramdump_size--;
+	rd_regionx_start[1] = rd_regionx_start[2];
+	rd_regionx_size[1] = rd_regionx_size[2];
 
-		if ((count % (16 * 1024)) == 0) {
-			// printf("%s: received %d bytes\n", __func__, count);
-			printf("\b=>");
-			fflush(stdout);
+	return 0;
+}
+
+static int send_region_info(int dev_fd, char *host_region)
+{
+	int ret;
+	char ack;
+
+	/* Send region info to TARGET */
+	ret = write(dev_fd, host_region, 1);
+	if (ret != 1) {
+		printf("Sending region info failed, ret = %d\n", ret);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int ramdump_recv(int dev_fd)
+{
+	char buf;
+	int i;
+	int ret;
+	int index = 1;
```

Comment:
if the num_of_regions are not more than 255, you can use unsigned char for i & index variable at multiple places used above. Hence multiple bytes can be saved to achieve space efficient code.